sklearn/tree/_classes.py

```diff
@@ -225,46 +230,65 @@ def fit(self, X, y, sample_weight=None, check_input=True):
             y = np.ascontiguousarray(y, dtype=DOUBLE)
 
         # Check parameters
+        if self.max_depth is not None:
+            check_scalar(
+                self.max_depth,
+                name="max_depth",
+                target_type=numbers.Integral,
+                min_val=1,
+            )
         max_depth = np.iinfo(np.int32).max if self.max_depth is None else self.max_depth
-        max_leaf_nodes = -1 if self.max_leaf_nodes is None else self.max_leaf_nodes
 
         if isinstance(self.min_samples_leaf, numbers.Integral):
-            if not 1 <= self.min_samples_leaf:
-                raise ValueError(
-                    "min_samples_leaf must be at least 1 or in (0, 0.5], got %s"
-                    % self.min_samples_leaf
-                )
+            check_scalar(
+                self.min_samples_leaf,
+                name="min_samples_leaf",
+                target_type=numbers.Integral,
+                min_val=1,
+                max_val=n_samples,
```

Comment:
This is technically not backward compatible. The following works to form single node trees:

```python
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeClassifier
X, y = make_classification(random_state=0)
clf = DecisionTreeClassifier(random_state=0, min_samples_leaf=X.shape[0] + 1)

clf.fit(X, y)
```

I'm guessing this could still have some educational value. I think we should remove the upper bound:

```suggestion
```

WDYT @glemaitre ?

Reply:
Yes we can remove the upper bound limit here.